File: src/adventofcode2018/Day13.java
package adventofcode2018;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

import utils.Input;

/**
 * https://adventofcode.com/2018/day/13
 */
public class Day13 {

    static class Cart {
        static int dirs[][] = {
            {1, 0, 3, 1}, {0, 1, 2, 0}, {-1, 0, 1, 3}, {0, -1, 0, 2}
        };
        int x, y, dir, nextTurn = -1;
        Cart(int x, int y, int dir) {
            this.x = x;
            this.y = y;
            this.dir = dir;
        }
        void move(int path) {
            if (path == '+') {
                dir = (dir + nextTurn + 4) % 4;
                nextTurn = (nextTurn + 2) % 3 - 1;  // -1 -> 0 -> 1
            } else if (path == '/') {
                dir = dirs[dir][2];
            } else if (path == '\\') {
                dir = dirs[dir][3];
            }
            x += dirs[dir][0];
            y += dirs[dir][1];
        }
    }

    public static void main(String[] args) {
        int track[][] = new Input(2018, 13).getMap();
        int h = track.length;
        int w = track[0].length;
        List<Cart> carts = new ArrayList<>();
        for (int i=0; i<h; i++) {
            for (int j=0; j<w; j++) {
                int c = track[i][j];
                if (c == '>') carts.add(new Cart(j, i, 0));
                else if (c == 'v') carts.add(new Cart(j, i, 1));
                else if (c == '<') carts.add(new Cart(j, i, 2));
                else if (c == '^') carts.add(new Cart(j, i, 3));
            }
        }
        List<Cart> cartsToRemove = new ArrayList<>();
        race:
        for (;;) {
            carts.sort(Comparator.comparing(c -> c.y*1000 + c.x));
            for (Cart c : carts) {
                c.move(track[c.y][c.x]);
                for (Cart c2 : carts) {
                    if (c != c2 && c.x == c2.x && c.y == c2.y) {
                        if (cartsToRemove.isEmpty()) {
                            System.err.println(c.x + "," + c.y);
                        }
                        cartsToRemove.add(c);
                        cartsToRemove.add(c2);
                        break;
                    }
                }
            }
            carts.removeAll(cartsToRemove);
            if (carts.size() == 1) {
                System.err.println(carts.get(0).x + "," + carts.get(0).y);
                break race;
            }
        }
    }
}

File: src/Python/2018/Day13.py

from utils import Input, printResult

# https://adventofcode.com/2018/day/13

dirs = [[1, 0, 3, 1], [0, 1, 2, 0], [-1, 0, 1, 3], [0, -1, 0, 2]]

class Cart:

    def __init__(self, x, y, dir):
        self.x, self.y = x, y
        self.dir = dir
        self.nextTurn = -1

    def move(self, path):
        if path == "+":
            self.dir = (self.dir + self.nextTurn + 4) % 4
            self.nextTurn = (self.nextTurn + 2) % 3 - 1
        elif path == "/":
            self.dir = dirs[self.dir][2]
        elif path == "\\":
            self.dir = dirs[self.dir][3]
        self.x += dirs[self.dir][0]
        self.y += dirs[self.dir][1]

track = Input(2018, 13).lines()
carts = []
for i in range(len(track)):
    for j in range(len(track[0])):
        c = track[i][j]
        if c == ">": carts.append(Cart(j, i, 0))
        if c == "v": carts.append(Cart(j, i, 1))
        if c == "<": carts.append(Cart(j, i, 2))
        if c == "^": carts.append(Cart(j, i, 3))

cartsToRemove = []
while len(carts) > 1:
    carts.sort(key = lambda c: c.y * 1000 + c.x)
    for c in carts:
        c.move(track[c.y][c.x])
        for c2 in carts:
            if c2 is not c and [c.x, c.y] == [c2.x, c2.y]:
                cartsToRemove.extend([c, c2])
                break
    for c in cartsToRemove:
        if c in carts: carts.remove(c)

printResult(1, ("%d,%d" % (cartsToRemove[0].x, cartsToRemove[0].y)))
printResult(2, ("%d,%d" % (carts[0].x, carts[0].y)))
